cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java

```diff
@@ -356,4 +400,161 @@ private void saveKeyStores(KeyStore serverKeyStore, KeyStore clientKeyStore, Loc
       }
     }
   }
+
+  /**
+   * Initialize {@link RuntimeMonitor} to re-monitor programs that are already running when this service starts.
+   *
+   * @param startMillis time in milliseconds that this service starts. It is used such that it will only monitor
+   *                    program runs launched before starting of this service.
+   */
+  private void initializeRuntimeMonitors(long startMillis) {
+    int limit = cConf.getInt(Constants.RuntimeMonitor.INIT_BATCH_SIZE);
+    AtomicReference<ProgramRunId> lastProgramRunId = new AtomicReference<>();
+    RetryStrategy retryStrategy = RetryStrategies.fromConfiguration(cConf, "system.runtime.monitor.");
+
+    boolean completed = false;
+
+    try {
+      while (!completed) {
+        // Scan the dataset and update the controllers map. Retry on all exception.
+        completed = Retries.callWithRetries(() -> Transactionals.execute(transactional, context -> {
+          RemoteRuntimeDataset dataset = RemoteRuntimeDataset.create(context, datasetFramework);
+          List<Map.Entry<ProgramRunId, ProgramOptions>> scanResult = dataset.scan(limit, lastProgramRunId.get());
+          for (Map.Entry<ProgramRunId, ProgramOptions> entry : scanResult) {
+            ProgramRunId programRunId = entry.getKey();
+            ProgramOptions programOptions = entry.getValue();
+
+            if (RunIds.getTime(programRunId.getRun(), TimeUnit.MILLISECONDS) > startMillis) {
```

Comment:
Seems like we should set lastProgramRunId earlier. If we scan a batch containing only these types of rows, we'll infinitely loop.

Reply:
Right. good catch.